### DISCUSSION TITLE
Subject: Inquiry Regarding Error in Profiling Analysis

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#27418] [Q&A General] [answered]
Dear MOOSE Forum Members,

I hope this email finds you well. I am writing to seek assistance regarding an error I encountered while conducting a profiling analysis using MOOSE.

I utilized the command `MOOSE_PROFILE_BASE=run1_ mpiexec -np 35 /home/pw-moose/projects/moose/modules/phase_field/phase_field-oprof -i grain_growth_2D_graintracker.i` to run my simulation, which resulted in the generation of a series of `run1_*.prof` files. However, when attempting to analyze one of these files using the `pprof run1.prof` command, I encountered the following error:

```
Using remote profile at run1.prof.
Use of uninitialized value $line in substitution (s///) at /home/pw-moose/miniforge/envs/moose/bin/pprof line 3326.
http://run1.prof/pprof/symbol doesn't exist
```

I have referred to the instructions provided on the MOOSE website, particularly the section on profiling (https://mooseframework.inl.gov/application_development/profiling.html), for guidance on conducting the profiling analysis.

Could you please advise on the possible reasons for this error and how I can resolve it? Any insights or suggestions would be greatly appreciated.

Thank you very much for your assistance.

Best regards,
Wei

**Comment by GiudGiud** [comment, marked as answer]
Hello

It seems this error is returned when the profile does not exist as I read here
https://github.com/gperftools/gperftools/issues/417

this is consistent with this log `Using remote profile at run1.prof.` which indicates it did not find it locally

so can you please paste:
- the content of the folder where you use this pprof command
- the exact command you ran

Guillaume

**Reply by PengWei97** [reply]
OK,

First, my test directory is ~/projects/moose/modules/phase_field/examples/grain_growth,

After that, I ran the command `MOOSE_PROFILE_BASE=run1_ mpiexec -np 35 /home/pw-moose/projects/moose/modules/phase_field/phase_field-oprof -i grain_growth_2D_graintracker.i`. After the simulation is completed, the specific files in this path are as shown below Shown below,
![image](https://github.com/idaholab/moose/assets/56183775/0d2c8887-a73b-49f2-a815-8dbef5e106b1)

After that, I executed the command `pprof run1.prof` or `pprof run1_1.prof`, and the result showed the following error,
![image](https://github.com/idaholab/moose/assets/56183775/e8f2d186-e552-4b9f-96a4-40d637a8c512)

wei

**Reply by PengWei97** [reply]
Meanwhile, https://github.com/gperftools/gperftools/issues/1506 says that we need to specify `program` before we can specify the profile file. Currently, I don’t know how to set this `program` in moose.

**Reply by GiudGiud** [reply]
the command is wrong I think
` pprof run1.prof ` does not work, run1.prof does not exist
`pprof run1_1.prof` is also not liking your input. I think it wants all the files to be specified? Or another 'profile file' that it is complaining about you not specifying it

**Reply by PengWei97** [reply]
Hi @GiudGiud,


Thank you for your suggestions. I found them very useful. Indeed, by reviewing the `pprof --help` output, I confirmed the correct command usage as follows:

```bash
pprof -top ~/projocts/panda/panda-oprof run3_1.prof
```

This approach resolved the issue with the command execution. However, I've noticed that running `pprof -top` on the specified profiling data takes a long time, and I'm seeking ways to optimize or understand the underlying cause of this delay.

I have created a new issue #27467 on the forum to address this specific challenge. I would greatly appreciate any insights or suggestions you might have on how to reduce the execution time or any related experiences you could share.

Thank you in advance for your time and help.

Best regards,

Wei

**Reply by GiudGiud** [reply]
I just followed the pprof installation instructions from the moose website and it seems to work fine.
which simulation did you profile? can you do this one?
```
cd moose/modules/phase_field/examples/grain_growth
MOOSE_PROFILE_BASE=run1_ mpiexec -n 2 ../../phase_field-oprof -i grain_growth_2D_graintracker.i Executioner/num_steps=2
~/gperftools/pprof-symbolize ../../phase_field-oprof run1_0.prof run1_1.prof 
```

**Reply by PengWei97** [reply]
Yes, the command you provided works normally, it just takes a long time. The results are as follows:
```bash
(moose) pw-moose@pwmoose-PowerEdge-T640:~/projocts/moose/modules/phase_field/examples/grain_growth$ ~/gperftools/pprof-symbolize ../../phase_field-oprof run1_0.prof run1_1.prof 
Using local file ../../phase_field-oprof.
Using local file run1_1.prof.
Using local file run1_0.prof.
Fetching 2 profiles, Be patient...
Using local file run1_1.prof.
Using local file run1_0.prof.
Welcome to pprof!  For help, type 'help'.
(pprof) top 12
Total: 27466 samples
    1302   4.7%   4.7%     1567   5.7% libMesh::DofMap::build_constraint_matrix [clone .localalias]
     844   3.1%   7.8%     2026   7.4% MooseVariableData::computeValues
     554   2.0%   9.8%     1034   3.8% MooseMesh::minPeriodicVector
     412   1.5%  11.3%      412   1.5% 0x000072db5b58946c
     319   1.2%  12.5%      646   2.4% std::_Rb_tree::_M_lower_bound (inline)
     311   1.1%  13.6%      311   1.1% hypre_BoomerAMGRelaxHybridGaussSeidel_core
     298   1.1%  14.7%      298   1.1% libMesh::TypeVector::operator* (inline)
     298   1.1%  15.8%      298   1.1% std::vector::size (inline)
     294   1.1%  16.9%      294   1.1% std::_Rb_tree::_S_right (inline)
     235   0.9%  17.7%      235   0.9% MooseArray::operator[] (inline)
(pprof) 
```

**Reply by GiudGiud** [reply]
so you dont see your material taking most of the compute time? How many time steps did you run?

**Reply by PengWei97** [reply]
I feel that these should be the problems in #27307. In fact, this email is mainly about a technical problem encountered in the process of seeking the cause of #27307. I used the input file to run the calculation example that comes with moose, for the purpose of testing correct use of pprof.
